loki/visitors/pretty_visualize.py

```diff
@@ -0,0 +1,349 @@
+
+    This serves as base class for backends and provides a number of helpful
+    routines that ease implementing automatic recursion and line wrapping.
+    It is adapted from the Stringifier in "pprint.py". It doubles as a means
+    to produce a human readable graph representation of the IR, which is
+    useful for debugging purposes and first visualization.
+
+    Parameters
+    ----------
+    linewidth : int, optional
+        The line width limit after which to break a line.
+    symgen : str, optional
+        A function handle that accepts a :any:`pymbolic.primitives.Expression`
+        and produces a string representation for that.
+    """
+
+    def __init__(
+        self,
+        linewidth=40,
+        symgen=str,
+    ):
+        super().__init__()
+        self.linewidth = linewidth
+        self._symgen = symgen
+        self._id = 0
+        self._id_map = {}
+
+    @property
+    def symgen(self):
+        """
+        Formatter for expressions.
+        """
+        return self._symgen
+
+    def join_items(self, items, sep=", ", separable=True):
+        """
+        Concatenate a list of items into :any:`JoinableStringList`.
+
+        The return value can be passed to :meth:`format_line` or
+        :meth:`format_node` or converted to a string with `str`, using
+        the :any:`JoinableStringList` as an argument.
+        Upon expansion, lines will be wrapped automatically to stay within
+        the linewidth limit.
+
+        Parameters
+        ----------
+        items : list
+            The list of strings to be joined.
+        sep : str, optional
+            The separator to be inserted between items.
+        separable : bool, optional
+            Allow line breaks between individual :data:`items`.
+
+        Returns
+        -------
+        :any:`JoinableStringList`
+        """
+        return JoinableStringList(
+            items,
+            sep=sep,
+            width=self.linewidth,
+            cont="\n",
+            separable=separable,
+        )
+
+    def format_node(self, name, *items):
+        """
+        Default format for a node.
+
+        Creates a string of the form ``<name[, attribute, attribute, ...]>``.
+        """
+        content = ""
+        if items:
+            content = self.format_line("<", name, " ", self.join_items(items), ">")
+        else:
+            content = self.format_line("<", name, ">")
+
+        # disregard all quotes to ensure nice graphviz label behaviour
+        return content.replace('"', "")
+
+    def format_line(self, *items, comment=None, no_wrap=False):
+        """
+        Format a line by concatenating all items and applying indentation while observing
+        the allowed line width limit.
+
+        Note that the provided comment will simply be extended to the line and no line
+        width limit will be enforced for that.
+
+        :param list items: the items to be put on that line.
```

Comment:
This uses a different docstring flavour, presumably because the Stringifier method had that? Would you mind aligning this with the style in the other methods?

Reply:
Fixed 37ee8f07da7f7bb1e804f46039de50e17e6d2d23